util/testutil/tar.go

```diff
@@ -493,3 +493,132 @@ func permAndExtraMode2TarMode(fm os.FileMode) (tm int64) {
+	return func(te *TarEntriesOptions) {
+		te.maxChildren = m
+	}
+}
+
+func WithRegProbability(p float32) TarEntriesOption {
+	return func(te *TarEntriesOptions) {
+		te.regProbability = p
+	}
+}
+
+func parseEntriesOptions(n int, te *TarEntriesOptions) error {
+	switch {
+	case n < 0:
+		return fmt.Errorf("entry count must not be less than 0")
+	case te.maxChildren < 0:
+		return fmt.Errorf("max children must not be less than 0")
+	case te.maxChildren > n:
+		return fmt.Errorf("cannot have more children than the required entry count")
+	}
+
+	if te.regProbability < 0 {
+		te.regProbability = 0
+	} else if te.regProbability == 0 {
+		te.regProbability = 0.5
+	}
+	if te.maxChildren == 0 {
+		te.maxChildren = (n / 2) + 1
+	}
+	return nil
+}
+
+// GenerateTarEntries generates a random set of tar entries of some given size. It creates
+// an m-ary tree of file/directory nodes and corresponding TarEntries for each.
+func GenerateTarEntries(entriesCount int, opts ...TarEntriesOption) ([]TarEntry, error) {
+	te := &TarEntriesOptions{}
+	for _, opt := range opts {
+		opt(te)
+	}
+	if err := parseEntriesOptions(entriesCount, te); err != nil {
+		return nil, err
+	}
+
+	var tarEntries []TarEntry
+
+	rootNode := &node{path: "", leaf: false}
+	nodeQueue := make([]*node, entriesCount+1)
+	nodeQueue[0] = rootNode
+	insertionIndex := 0
+	for insertionIndex < len(nodeQueue)-1 {
+		// POP() the first item in the queue
+		nd := nodeQueue[0]
+		nodeQueue = nodeQueue[1:]
+		// If the node is a leaf, continue.
+		if nd.isLeaf() {
+			insertionIndex = insertionIndex - 1
+			continue
+		}
+		remainingEmptyNodes := len(nodeQueue) - insertionIndex
+		// Get children count [1,maxChildren]. The children count must not
+		// exceed the remaining empty nodes, so we take the min.
+		m := min(randRangeInc(1, te.maxChildren), remainingEmptyNodes)
+		endIndex := insertionIndex + m
+		nonLeafIndex := randRangeInc(insertionIndex, endIndex-1)
+
+		// Starting from the insertionIndex, create m children and add them to the queue.
+		for k := insertionIndex; k < endIndex; k++ {
+			childNode := &node{}
+			// At least 1 child node needs to a non-leaf, unless it is the last node in the tree.
+			if k == nonLeafIndex && k != len(nodeQueue)-1 {
+				childNode.leaf = false
+			} else {
+				childNode.leaf = shouldBeLeaf(te.regProbability)
+			}
+			path := RandString(10)
+			if childNode.isLeaf() {
+				childNode.path = fmt.Sprintf("%s%s.txt", nd.path, path)
+				tarEntries = append(tarEntries, File(childNode.path, ""))
+			} else {
+				childNode.path = fmt.Sprintf("%s%s/", nd.path, path)
+				tarEntries = append(tarEntries, Dir(childNode.path))
+			}
+			// ENQUEUE() childNode
+			nodeQueue[k] = childNode
+		}
+		insertionIndex = (insertionIndex - 1) + m
+	}
+
+	return tarEntries, nil
+}
+
+// randRangeInc returns a random range between [min,max] inclusive.
+func randRangeInc(min int, max int) int {
+	if min == max {
+		return min
+	}
+	return r.Intn((max+1)-min) + min
```

Comment:
I probably should have documented this better, but the intention with having this unexported, global `ThreadSafeRandom` was just to make it so that we didn't have to go back and update every existing test immediately.

I don't think any new code should use it since it makes reproducibility dependent on test execution order. The idea was that we want "randomness" in the sense that we want various input object structures, but we actually want comparability by making the test deterministic.

Reply:
> I don't think any new code should use it since it makes reproducibility dependent on test execution order. The idea was that we want "randomness" in the sense that we want various input object structures, but we actually want comparability by making the test deterministic.

Sorry, does the seeded random not return the same value regardless of execution order and whether or not we use the global or a local instance?

Reply:
The TSR will generate the same random sequence, but if multiple tests use the same TSR, then the sequence they see depends on the order they run.


If you create a new TSR in a test and only use that, then the input data will be exactly the same every single time you run the test, not matter what other tests are running. If you use the global TSR, then the input depends on what other tests are running and in what order.